### PR TITLE
[Model Monitoring] add create_model_monitoring_alert_configs to project.

### PR DESCRIPTION
**Pull request by tomerm-iguazio**
1. Add regex methods for result_names.
2. Finished `create_model_monitoring_alert_configs`
3. Created 3 test:
    a) Unit test in `test_helpers.py`, in order to check regex functionallity.
    b) integration test in `test_monitoring_api.py`, with 2 MEPS and different metrics, with mocking the db.
    c) Full system test that added to `test_drift_detection_alert` in `test_alerts.py`.

[ML-8477](https://iguazio.atlassian.net/browse/ML-8477)

[ML-8477]: https://iguazio.atlassian.net/browse/ML-8477?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

File: tests/system/model_monitoring/test_model_monitoring.py [hunk 1]
```diff
@@ -330,9 +332,10 @@ def test_basic_model_monitoring(self, with_sql_target: bool) -> None:
             )
             sleep(choice([0.01, 0.04]))
 
-        sleep(5)
+        sleep(15)
         endpoints_list = mlrun.get_run_db().list_model_endpoints(
```
Comment (tomerm-iguazio) [inline tests/system/model_monitoring/test_model_monitoring.py:336]: add metrics to `project. list_model_endpoints` so we will be able to use it here.

File: tests/system/model_monitoring/test_model_monitoring.py [hunk 2]
```diff
@@ -353,10 +356,30 @@ def test_basic_model_monitoring(self, with_sql_target: bool) -> None:
             self.project_name, endpoint.metadata.uid
         )
         assert len(metrics) == 1
-        assert (
-            metrics[0].full_name
-            == f"{self.project_name}.mlrun-infra.metric.invocations"
-        )
+        expected_metric_name = f"{endpoint.metadata.uid}.mlrun-infra.metric.invocations"
+        assert metrics[0].full_name == expected_metric_name
```
Comment (tomerm-iguazio) [inline tests/system/model_monitoring/test_model_monitoring.py]: fix to fqn

File: mlrun/projects/project.py [hunk 3]
```diff
@@ -3480,6 +3542,7 @@ def list_model_endpoints(
         labels: Optional[list[str]] = None,
         start: str = "now-1h",
         end: str = "now",
+        metrics: Optional[list[str]] = None,
```
Comment (davesh0812) [inline mlrun/projects/project.py]: This can be work only with v3io tsdb, we are changing `list_model_endpoints` in #6796.

Why did you add this here? Note that it is not related to the application metrics.

Comment (tomerm-iguazio) [reply mlrun/projects/project.py]: removed

File: mlrun/projects/project.py [hunk 4]
```diff
@@ -2040,6 +2045,63 @@ def setup(self, save: bool = True) -> "MlrunProject":
...
+        db = mlrun.db.get_run_db(secrets=self._secrets)
+        metrics = []
+        alerts = []
+        # TODO: Refactor to use a single request to improve performance at scale, ML-8473
+        for endpoint in endpoints:
+            metrics_by_endpoint = db.get_model_endpoint_monitoring_metrics(
+                project=self.name, endpoint_id=endpoint.metadata.uid
+            )
+            metrics_fqn_by_endpoint = [
+                get_result_instance_fqn(
```
Comment (davesh0812) [inline mlrun/projects/project.py:2085]: I don't think you need this, the full name is already part of metric/result.

File: mlrun/projects/project.py [hunk 5]
```diff
@@ -2040,6 +2045,63 @@ def setup(self, save: bool = True) -> "MlrunProject":
...
+        notifications: list[alert_constants.AlertNotification],
+        result_names: Optional[
+            list[str]
+        ] = None,  # can use wildcards - see below for explanation.
+        severity: alert_constants.AlertSeverity = alert_constants.AlertSeverity.MEDIUM,
+        criteria: alert_constants.AlertCriteria = alert_constants.AlertCriteria(
+            count=1, period="10m"
+        ),
+        reset_policy: mlrun.common.schemas.alert.ResetPolicy = mlrun.common.schemas.alert.ResetPolicy.AUTO,
+    ) -> list[mlrun.alerts.alert.AlertConfig]:
```
Comment (davesh0812) [inline mlrun/projects/project.py:2054]: please add documentation

File: tests/system/model_monitoring/test_model_monitoring.py [hunk 6]
```diff
@@ -330,9 +333,10 @@ def test_basic_model_monitoring(self, with_sql_target: bool) -> None:
             )
             sleep(choice([0.01, 0.04]))
 
-        sleep(5)
+        sleep(15)
         endpoints_list = mlrun.get_run_db().list_model_endpoints(
-            self.project_name, metrics=["predictions_per_second"]
+            self.project_name,
+            metrics=[mm_constants.EventLiveStats.PREDICTIONS_PER_SECOND],
```
Comment (davesh0812) [inline tests/system/model_monitoring/test_model_monitoring.py]: delete this as well

**Review by davesh0812 — COMMENTED** [review]
Looks good, few comments.

Also can you add this flow to the `test_drift_detection_alert`?

File: mlrun/model_monitoring/helpers.py [hunk 7]
```diff
@@ -45,6 +48,54 @@ class _BatchDict(typing.TypedDict):
...
+    for result_name in result_names:
+        if fnmatch(metric_name, result_name):
+            return True
+    return False
+
+
+def filter_metrics_by_regex(
+    metrics_names: Optional[list[str]] = None,
+    result_names: Optional[list[str]] = None,
+) -> list[str]:
```
Comment (davesh0812) [inline mlrun/model_monitoring/helpers.py:73]: docs

File: mlrun/model_monitoring/helpers.py [hunk 8]
```diff
@@ -45,6 +48,54 @@ class _BatchDict(typing.TypedDict):
...
+    metric_name = ".".join(metric_name.split(".")[i] for i in [1, 3])
+    for result_name in result_names:
+        if fnmatch(metric_name, result_name):
+            return True
+    return False
+
+
+def filter_metrics_by_regex(
+    metrics_names: Optional[list[str]] = None,
+    result_names: Optional[list[str]] = None,
```
Comment (davesh0812) [inline mlrun/model_monitoring/helpers.py]: rename parmas

File: mlrun/projects/project.py [hunk 9]
```diff
@@ -2040,6 +2045,63 @@ def setup(self, save: bool = True) -> "MlrunProject":
...
+        db = mlrun.db.get_run_db(secrets=self._secrets)
+        metrics = []
+        alerts = []
+        # TODO: Refactor to use a single request to improve performance at scale, ML-8473
+        for endpoint in endpoints:
+            metrics_by_endpoint = db.get_model_endpoint_monitoring_metrics(
+                project=self.name, endpoint_id=endpoint.metadata.uid
```
Comment (davesh0812) [inline mlrun/projects/project.py]: specify the type to result.

File: mlrun/model_monitoring/helpers.py [hunk 10]
```diff
@@ -45,6 +48,54 @@ class _BatchDict(typing.TypedDict):
...
+                f"filter_metrics_by_regex: result_name illegal, will be ignored."
+                f" Result_name: {result_name}"
+            )
+        else:
+            validated_result_names.append(result_name)
+    filtered_metrics_names = []
+    for metric_name in metrics_names:
+        if metric_name.count(".") != 3 or any(
+            part == "" for part in metric_name.split(".")
+        ):
```
Comment (theSaarco) [inline mlrun/model_monitoring/helpers.py]: Why do it here? This seems to be logic that should be verified by the `_is_metrics_regex_match` function.

File: mlrun/model_monitoring/helpers.py [hunk 11]
```diff
@@ -45,6 +48,54 @@ class _BatchDict(typing.TypedDict):
     days: int
 
 
+def _is_metrics_regex_match(
+    metric_name: Optional[str],
+    result_names: Optional[list[str]],
+):
+    metric_name = ".".join(metric_name.split(".")[i] for i in [1, 3])
+    for result_name in result_names:
+        if fnmatch(metric_name, result_name):
```
Comment (theSaarco) [inline mlrun/model_monitoring/helpers.py]: Are the names of metrics case-sensitive? If they are then you should be using `fnmatchcase()` instead.

**Review by theSaarco — COMMENTED** [review]
Looks good - some minor comments.

File: mlrun/projects/project.py (hunk 4 above)
Comment (tomerm-iguazio) [reply mlrun/projects/project.py:2085]: full name and fqn is not identical:

**full name** starting with `project.name` - for example "my_project.my_app.result.my_result" - as described in `ModelEndpointMonitoringMetric._compose_full_name`

**fqn**  starting with `model_endpoint.metadata.uid` for example "my_endpoint.my_app.result.my_result" - as described in `get_result_instance_fqn`

File: tests/system/model_monitoring/test_model_monitoring.py (hunk 6 above)
Comment (tomerm-iguazio) [reply tests/system/model_monitoring/test_model_monitoring.py]: why?

File: mlrun/model_monitoring/helpers.py (hunk 10 above)
Comment (tomerm-iguazio) [reply mlrun/model_monitoring/helpers.py]: moved, tnx!

File: mlrun/model_monitoring/helpers.py (hunk 8 above)
Comment (tomerm-iguazio) [reply mlrun/model_monitoring/helpers.py:64]: fix

File: tests/system/model_monitoring/test_model_monitoring.py (hunk 6 above)
Comment (davesh0812) [reply tests/system/model_monitoring/test_model_monitoring.py]: there is no `metrics` param in the new version of `list_model_endpoints()` method

**Review by davesh0812 — APPROVED** [review]
Nice!

**Review by assaf758 — APPROVED** [review]
